crates/typst/src/foundations/calc.rs

```diff
@@ -660,44 +709,57 @@ pub fn trunc(
 ///
 /// ```example
 /// #assert(calc.fract(3) == 0)
+/// #assert(calc.fract(decimal("234.23949211")) == decimal("0.23949211"))
 /// #calc.fract(-3.1)
 /// ```
 #[func(title = "Fractional")]
 pub fn fract(
     /// The number to truncate.
-    value: Num,
-) -> Num {
+    value: DecNum,
+) -> DecNum {
     match value {
-        Num::Int(_) => Num::Int(0),
-        Num::Float(n) => Num::Float(n.fract()),
+        DecNum::Decimal(n) => DecNum::Decimal(n.fract()),
+        DecNum::Int(_) => DecNum::Int(0),
+        DecNum::Float(n) => DecNum::Float(n.fract()),
     }
 }
 
 /// Rounds a number to the nearest integer.
 ///
 /// Optionally, a number of decimal places can be specified.
 ///
+/// Note that this function will return the same type as the operand.
+/// That is, applying `round` to a [`{float}`]($float) will return a `{float}`,
+/// and to a [`{decimal}`]($decimal), another `{decimal}`. You may explicitly
+/// convert the output of this function to an integer with [`{int}`]($int), but
+/// note that such a conversion will error if the `{float}` or `{decimal}` is
+/// larger than the maximum 64-bit signed integer or smaller than the minimum
+/// integer.
+///
 /// ```example
+/// #assert(calc.round(3) == 3)
 /// #assert(calc.round(3.14) == 3)
-/// #assert(calc.round(3.5) == 4)
+/// #assert(calc.round(3.5) == 4.0)
+/// #assert(calc.round(decimal("-6.5")) == decimal("-7"))
+/// #assert(calc.round(decimal("7.123456789"), digits: 6) == decimal("7.123457"))
 /// #calc.round(3.1415, digits: 2)
 /// ```
 #[func]
 pub fn round(
     /// The number to round.
-    value: Num,
-    /// The number of decimal places.
+    value: DecNum,
+    /// The number of decimal places. Must not be negative.
     #[named]
     #[default(0)]
-    digits: i64,
-) -> Num {
+    digits: u32,
```

Comment:
I feel like this is the best way forward. It greatly simplifies the function and removes previously inconsistent behavior.

(This felt noteworthy to add a comment about, in case any further discussion is needed.)